src/exchanges/bitfinex-client.js

```diff
@@ -207,9 +234,79 @@ class BitfinexClient extends BasicClient {
   }
 
   _onTradeMessage(msg, market) {
-    // example msg: [ 359491, 'tu', [ 560287312, 1609712228656, 0.005, 33432 ], 6 ]
-    let [id, unix, amount, price] = msg[2];
+    const timestampMs = msg[3];
+    if (msg[1] === "hb") {
+      const sequenceId = Number(msg[2]);
+      if (this.enableEmptyHeartbeatEvents === false) return;
+      // handle heartbeat by emitting empty update w/sequenceId.
+      // example trade heartbeat msg: [ 198655, 'hb', 3, 1610920929093 ]
+      let trade = new Trade({
+        exchange: "Bitfinex",
+        base: market.base,
+        quote: market.quote,
+        timestamp: timestampMs,
+        sequenceId,
+      });
+      this.emit("trade", trade, market);
+      return;
+    }
+    if (Array.isArray(msg[1])) {
```

Comment:
Suggest extracting to `_onTradeHistory` or appropriately named method and adding method comment with method JSON (might not need to include whole thing if it's huge, just enough to get the gist.